akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala

```diff
@@ -139,32 +225,80 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with LogHelper
         case x => receivedUnexpectedFrame(x)
       }
 
+    def pullNextFrame(maxSize: Int): (StreamState, PullFrameResult) = throw new IllegalStateException(s"pullNextFrame not supported in state $stateName")
+
     /** Called to cleanup any state when the connection is torn down */
-    def shutdown(): IncomingStreamState
+    def shutdown(): Unit = ()
   }
-  case object Idle extends IncomingStreamState {
-    def handle(event: StreamFrameEvent): IncomingStreamState = expectIncomingStream(event, HalfClosedRemote, ReceivingDataFirst)
-    override def handleOutgoingCreated(stream: Http2SubStream): IncomingStreamState = SendingData(stream)
-    override def shutdown(): IncomingStreamState = Closed
+
+  case object Idle extends StreamState {
+    def handle(event: StreamFrameEvent): StreamState = expectIncomingStream(event, HalfClosedRemoteWaitingForOutgoingStream(0), OpenReceivingDataFirst(_, 0))
+    override def handleOutgoingCreated(outStream: OutStream, correlationAttributes: Map[AttributeKey[_], _]): StreamState = OpenSendingData(outStream, correlationAttributes)
+    override def handleOutgoingCreatedAndFinished(correlationAttributes: Map[AttributeKey[_], _]): StreamState = HalfClosedLocalWaitingForPeerStream(correlationAttributes)
+  }
+  case class OpenReceivingDataFirst(buffer: IncomingStreamBuffer, extraInitialWindow: Int = 0) extends ReceivingData(HalfClosedRemoteWaitingForOutgoingStream(extraInitialWindow)) {
+    override def handleOutgoingCreated(outStream: OutStream, correlationAttributes: Map[AttributeKey[_], _]): StreamState = {
+      outStream.increaseWindow(extraInitialWindow)
+      Open(buffer, outStream)
+    }
+    override def handleOutgoingCreatedAndFinished(correlationAttributes: Map[AttributeKey[_], _]): StreamState = HalfClosedLocal(buffer)
+    override def handleOutgoingEnded(): StreamState = Closed
+
+    override protected def onReset(streamId: Int): Unit = multiplexer.closeStream(streamId)
+
+    override def incrementWindow(delta: Int): StreamState = copy(extraInitialWindow = extraInitialWindow + delta)
   }
-  case class ReceivingDataFirst(buffer: IncomingStreamBuffer) extends ReceivingData(HalfClosedRemote) {
-    override def handleOutgoingCreated(stream: Http2SubStream): IncomingStreamState = Open(buffer)
-    override def handleOutgoingEnded(): IncomingStreamState = Closed
+  trait Sending extends StreamState { _: Product =>
+    protected def outStream: OutStream
+
+    override def pullNextFrame(maxSize: Int): (StreamState, PullFrameResult) = {
+      val frame = outStream.nextFrame(maxSize)
```

Comment:
Here the old and somewhat unintuitive `nextFrame` / `endStreamIfPossible` interface is still existent and converted into the nicer `PullFrameResult` one. It's now internal to stream handling so less critical to remove (but might be removed in another PR).

Reply:
Created https://github.com/akka/akka-http/issues/3575 to track that work